pkg/cmd/cli/debug/debug.go

```diff
@@ -0,0 +1,140 @@
+	kubeconfigPath string
+	// optional, the name of the backup resource whose log will be packaged into the debug bundle
+	backup string
+	// optional, the name of the restore resource whose log will be packaged into the debug bundle
+	restore string
+}
+
+func (o *option) bindFlags(flags *pflag.FlagSet) {
+	flags.StringVar(&o.outputPath, "output", "", "The path of the bundle tarball, by default it's $HOME/bundle.tar.gz. Optional")
+	flags.StringVar(&o.backup, "backup", "", "The name of the backup resource whose log will be collected, no backup logs will be collected if it's not set. Optional")
+	flags.StringVar(&o.restore, "restore", "", "The name of the restore resource whose log will be collected, no restore logs will be collected if it's not set. Optional")
+}
+
+func (o *option) asCrashdArgs() string {
+	return fmt.Sprintf("output=%s,namespace=%s,basedir=%s,backup=%s,restore=%s,kubeconfig=%s",
+		o.outputPath, o.namespace, o.baseDir, o.backup, o.restore, o.kubeconfigPath)
+}
+
+func (o *option) complete(f client.Factory, fs *pflag.FlagSet) error {
+	if len(o.outputPath) > 0 {
+		absOutputPath, err := filepath.Abs(o.outputPath)
+		if err != nil {
+			return fmt.Errorf("invalid output path: %v", err)
+		}
+		o.outputPath = absOutputPath
+	}
+	tmpDir, err := ioutil.TempDir("", "crashd")
+	if err != nil {
+		return err
+	}
+	o.baseDir = tmpDir
+	o.namespace = f.Namespace()
+	kp := kubeconfig(fs)
+	o.kubeconfigPath, err = filepath.Abs(kp)
+	if err != nil {
+		return fmt.Errorf("invalid kubeconfig path: %s, %v", kp, err)
+	}
+	return nil
+}
+
+// NewCommand creates a cobra command.
+func NewCommand(f client.Factory) *cobra.Command {
+	o := &option{}
+	c := &cobra.Command{
+		Use:   "debug",
+		Short: "Generate debug bundle",
+		Long: `Generate a tarball containing the logs of velero deployment, plugin logs, restic DaemonSet, 
+specs of BackupStorageLocations, PodVolumeBackups, PodVolumeRestores, and optionally the specs and logs of backup and restore.`,
+		Run: func(c *cobra.Command, args []string) {
+			defer func(opt *option) {
+				if len(o.baseDir) > 0 {
+					if err := os.RemoveAll(o.baseDir); err != nil {
+						fmt.Fprintf(os.Stderr, "Failed to remove temp dir: %s: %v\n", o.baseDir, err)
+					}
+				}
+			}(o)
+			flags := c.Flags()
+			err := o.complete(f, flags)
+			cmd.CheckError(err)
+			err2 := runCrashd(o.asCrashdArgs())
+			cmd.CheckError(err2)
+		},
+	}
+	o.bindFlags(c.Flags())
+	return c
+}
+
+func runCrashd(argString string) error {
+	bak := os.Args
+	defer func() { os.Args = bak }()
+	f, err := ioutil.TempFile("", "velero*.cshd")
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := os.Remove(f.Name()); err != nil {
+			fmt.Fprintf(os.Stderr, "Failed to remove the temp file: %s, %v\n", f.Name(), err)
+		}
+	}()
+	_, err2 := f.Write(scriptBytes)
+	if err2 != nil {
+		return err2
+	}
+	os.Args = []string{"", "run", "--debug", f.Name(), "--args", fmt.Sprintf("%s", argString)}
+	return crashdCmd.Run()
```

Comment:
Did you consider using the [`ExecuteFile`](https://github.com/vmware-tanzu/crash-diagnostics/blob/main/exec/executor.go#L36-L38) function here? That way we might be able to avoid writing to `os.Args` and can pass in a map of arguments.

Reply:
Thanks for pointing it out, I updated the code to use `exec.Execute` which consumes an `io.Reader` so there's no need to write the script to a temp file.